linkerd/protocol/http/src/main/scala/io/buoyant/linkerd/protocol/ZipkinTracePropagator.scala

```diff
@@ -61,28 +59,122 @@ object ZipkinTrace {
+            case 0 =>
+              // bogus
+              ()
+            case 1 =>
+              // either debug flag or sampled
+              handleSampled(headers, a(0))
+            case 2 =>
+              // this is required to be traceId, spanId
+              handleTraceAndSpanIds(headers, a(0), a(1))
+            case 3 =>
+              handleTraceAndSpanIds(headers, a(0), a(1))
+              handleSampled(headers, a(2))
+            case 4 =>
+              handleTraceAndSpanIds(headers, a(0), a(1))
+              handleSampled(headers, a(2))
+              headers.set(ZipkinParentHeader, a(3))
+          }
+        case None =>
+          ()
+      }
+      headers -= ZipkinB3SingleHeader
+    }
+  }
+
+  /**
+   * Get a trace id from the request, if it has one, in either the historical multiple "x-b3-" headers or the
+   * newer "b3" single header. The "b3" single-header variant takes precedence over the multiple header one
+   * when extracting fields, that also implies ignoring the latter if "b3" single header is read.
+   * There's no way to know what the downstream is capable of, and it is more likely it supports "X-B3-*" vs not, so,
+   * the portable choice is to always write down "X-B3-"  even if we read "b3". Finagle does a similar thing in
+   * finagle-http-base: read "b3", write back "X-B3-".
+   * @note The "b3" encodings specs, single or multi: https://github.com/openzipkin/b3-propagation#http-encodings
+   * @param headers
+   * @return Option[TraceId]
+   */
   def get(headers: HeaderMap): Option[TraceId] = {
-    val trace = caseInsensitiveGet(headers, ZipkinTraceHeader).flatMap(SpanId.fromString)
+    //if "b3" present convert it to "x-b3-" multio and let the flow continue as if it has received "x-b3-"
+    convertB3SingleHeaderToMultiHeaders(headers)
+
+    // expect to read a 128bit traceid field, b3 single header supports 128bit traceids
+    val trace128Bit = caseInsensitiveGet(headers, ZipkinTraceHeader) match {
+      case Some(s) => TraceId128(s)
+      case None => TraceId128.empty
+    }
+
     val parent = caseInsensitiveGet(headers, ZipkinParentHeader).flatMap(SpanId.fromString)
     val span = caseInsensitiveGet(headers, ZipkinSpanHeader).flatMap(SpanId.fromString)
     val sample = caseInsensitiveGet(headers, ZipkinSampleHeader).map(StringUtil.toBoolean)
     val flags = caseInsensitiveGet(headers, ZipkinFlagsHeader).map(StringUtil.toSomeLong) match {
       case Some(f) => Flags(f)
       case None => Flags()
     }
+
     span.map { s =>
-      TraceId(trace, parent, s, sample, flags)
+      TraceId(trace128Bit.low, parent, s, sample, flags, trace128Bit.high)
     }
   }
 
   def set(headers: HeaderMap, id: TraceId): Unit = {
     val _ = headers.set(ZipkinSpanHeader, id.spanId.toString)
-    val __ = headers.set(ZipkinTraceHeader, id.traceId.toString)
+
+    // support setting a 128bit traceid
+    if (id.traceIdHigh.isEmpty) {
+      val __ = headers.set(ZipkinTraceHeader, id.traceId.toString)
```

Comment:
no need for assignign this to a val `___`

Reply:
Done, removed all assignments.